homeassistant/components/squeezebox/coordinator.py

```diff
@@ -0,0 +1,86 @@
+    STATUS_SENSOR_NEWVERSION,
+    STATUS_SENSOR_RESCAN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class LMSStatusDataUpdateCoordinator(DataUpdateCoordinator):
+    """LMS Status custom coordinator."""
+
+    def __init__(self, hass: HomeAssistant, my_api: Server) -> None:
+        """Initialize my coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=my_api.name,
+            update_interval=timedelta(seconds=SENSOR_UPDATE_INTERVAL),
+            always_update=False,
+        )
+        self.my_api = my_api
+        self.newversion_regex = re.compile("<.*$")
+
+    async def _async_update_data(self):
+        """Fetch data fromn LMS status call.
+
+        Then we process only a subset to make then nice for HA
+        """
+        # Note: asyncio.TimeoutError and aiohttp.ClientError are already
+        # handled by the data update coordinator.
+        async with timeout(STATUS_API_TIMEOUT):
+            data = await self.my_api.async_status()
+
+        if not data:
+            raise UpdateFailed("No data from status poll")
+
+        _LOGGER.debug("Raw serverstatus %s=%s", self.my_api.name, data)
+        # Sensor that need special handling
+        # 'lastscan': '1718431678', epoc -> ISO 8601 not always present
+        data[STATUS_SENSOR_LASTSCAN] = (
+            STATUS_SENSOR_LASTSCAN in data
+            and dt_util.utc_from_timestamp(int(data[STATUS_SENSOR_LASTSCAN]))
+            or None
+        )
+        # rescan # bool are we rescanning alter poll not always present
+        data[STATUS_SENSOR_RESCAN] = STATUS_SENSOR_RESCAN in data and True or False
+        # needsrestart bool plugin updates... not always present
+        data[STATUS_SENSOR_NEEDSRESTART] = (
+            STATUS_SENSOR_NEEDSRESTART in data and True or False
+        )
+        # newversion str not always present
+        # Sample text 'A new version of Logitech Media Server is available (8.5.2 - 0). <a href="updateinfo.html?installerFile=/var/lib/squeezeboxserver/cache/updates/logitechmediaserver_8.5.2_amd64.deb" target="update">Click here for further information</a>.'
+        data[STATUS_SENSOR_NEWVERSION] = (
+            STATUS_SENSOR_NEWVERSION in data
+            and self.newversion_regex.sub("...", data[STATUS_SENSOR_NEWVERSION])
+            or None
+        )
+        # newplugins str not always present
+        data[STATUS_SENSOR_NEWPLUGINS] = (
+            STATUS_SENSOR_NEWPLUGINS in data and data[STATUS_SENSOR_NEWPLUGINS] or None
+        )
+
+        # "info total duration"  in fractions of seconds
+        # progressdone	Returned with the current value of items completed for current scan phase. Not returned if no scan is in progress.
+        # progresstotal	Returned with the total value of items found for current scan phase. Not returned if no scan is in progress.
+        # lastscanfailed Information about a possible failure in case a scan has not finished in an attended manner.
```

Comment:
Why isn't this in the library? We can remove a lot of the comments and make it much simpler

Reply:
Surely the data presentation should be down to the integration? , The library does I would expect in this case and gives me the results of my query.

Reply:
You could , and you could argue this is good defensive programming and response should be perceived by the presentation layer not hidden by an intervening api as we might be able to use that subtlety to better present the data.  I have been bitten by the curse of the API hiding that to many times.

Reply:
comprise take it out of the _async_update_data and put it the class in say _prepare_data to separate out the Update and data manipulation?

Reply:
I mean, i dont mind it being here, but we should not make it too complicated with alle comments and stuff. We should rather introduce unit tests to make sure the data is perceived correctly

Reply:
I think this code would still need to be there even with unit tests but I can take out the comments about the status Items I haven't used or manipulated for presentation, they were more an Aid for later.

Reply:
I think it's nice to have unit tests so we know what data we expect and how the entities we create represent them, so we also validate this code

Reply:
agreed but I am not currently able to write that code(I haven't got to grips with the testing framework and I had a quick look at the some tests and ran away screaming) and it might be a few weeks, I still have the senors file to go can we look at the tests then? as at least that reduces the uncommitted state to that last bit of work.

Reply:
OK I have had a go at the tests